controllers/argocd/applicationset.go

```diff
@@ -308,34 +358,436 @@ func (r *ReconcileArgoCD) reconcileApplicationSetServiceAccount(cr *argoproj.Arg
+		}
+		return r.Client.Create(context.TODO(), clusterRB)
+	}
+
+	// ArgoCD not cluster scoped, cleanup any existing resource and exit
+	if !allowed {
+		err := r.Client.Delete(context.TODO(), existingClusterRB)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+		}
+		return nil
+	}
+
+	// if subj differ, update the rolebinding
+	if !reflect.DeepEqual(existingClusterRB.Subjects, clusterRB.Subjects) {
+		existingClusterRB.Subjects = clusterRB.Subjects
+		if err := r.Client.Update(context.TODO(), existingClusterRB); err != nil {
+			return err
+		}
+	} else if !reflect.DeepEqual(existingClusterRB.RoleRef, clusterRB.RoleRef) {
+		// RoleRef can't be updated, delete the rolebinding so that it gets recreated
+		r.Client.Delete(context.TODO(), existingClusterRB)
+		return fmt.Errorf("change detected in roleRef for rolebinding %s of Argo CD instance %s in namespace %s", existingClusterRB.Name, cr.Name, existingClusterRB.Namespace)
+	}
+	return nil
+}
+
+func (r *ReconcileArgoCD) reconcileApplicationSetSourceNamespacesResources(cr *argoproj.ArgoCD) error {
+
+	var reconciliationErrors []error
+
+	// controller disabled, nothing to do. cleanup handled by removeUnmanagedApplicationSetSourceNamespaceResources()
+	if cr.Spec.ApplicationSet == nil {
+		return nil
+	}
+
+	// create resources for each appset source namespace
+	for _, sourceNamespace := range cr.Spec.ApplicationSet.SourceNamespaces {
+
+		// skip source ns if doesn't exist
+		namespace := &corev1.Namespace{}
+		if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: sourceNamespace}, namespace); err != nil {
+			errMsg := fmt.Errorf("failed to retrieve namespace %s", sourceNamespace)
+			reconciliationErrors = append(reconciliationErrors, errors.Join(errMsg, err))
+			continue
+		}
+
+		// No namespace can be managed by multiple argo-cd instances (cluster scoped or namespace scoped)
+		// i.e, only one of either managed-by or applicationset-managed-by-cluster-argocd labels can be applied to a given namespace.
+		// Since appset-in-any-ns is in beta, we prioritize managed-by label in case of a conflict.
+		if value, ok := namespace.Labels[common.ArgoCDManagedByLabel]; ok && value != "" {
+			log.Info(fmt.Sprintf("Skipping reconciling resources for namespace %s as it is already managed-by namespace %s.", namespace.Name, value))
+			// remove any source namespace resources
+			if val, ok1 := namespace.Labels[common.ArgoCDApplicationSetManagedByClusterArgoCDLabel]; ok1 && val != cr.Namespace {
+				delete(r.ManagedApplicationSetSourceNamespaces, namespace.Name)
+				if err := r.cleanupUnmanagedApplicationSetSourceNamespaceResources(cr, namespace.Name); err != nil {
+					log.Error(err, fmt.Sprintf("error cleaning up resources for namespace %s", namespace.Name))
+				}
+			}
+			continue
+		}
+
+		log.Info(fmt.Sprintf("Reconciling applicationset resources for %s", namespace.Name))
+		// add applicationset-managed-by-cluster-argocd label on namespace
```

Comment:
@svghadi do we need a dedicated label for `appset-managed-by-cluster-argocd`? We know that for a namespace to be able to host an appset, it must already be included as an app source namespace. That means it already contains the app-source ns label, so no other instance will try to source apps from there 

I'm not sure I see the value of having an additional label to indicate that appsets in a given namespace are managed by xyz instance

Reply:
The new label is more for tracking appsets namespaces for role & rolebinding cleanup when there are no longer part of ArgoCD CR, but based off your previous comment, if we want to completely skip reconciliation when target appsets ns is not part of apps-in-any-ns (which makes sense), I think we should be able to get rid of this new label.

Reply:
hmmm, on second thought I think the clean up aspect is a good point
even though we dont technically need this specific label for clean up when a ns is no longer an app source ns, its more intuitive and keeps the logic uniform going forwards

Plus, there is the case when the subset of appset namespaces changes within the set of source namespaces, so a namespace could still be a source namespace but not an appset namespace anymore, in which case we may want to take action in that namespace. If we just go by the calculated list of new namespaces (and dont compare against the list of labelled namespaces) we wont have a way to know which namespace was just removed from it

Reply:
As discussed offline, keeping the new label for cleanup purpose.